pkg/controller/namespace.go

```diff
@@ -0,0 +1,271 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/cert-manager/istio-csr/cmd/app/options"
+)
+
+const (
+	resyncPeriod        = time.Second * 60
+	IstioConfigLabelKey = "istio.io/config"
+)
+
+// CARoot manages reconciles a configmap in each namespace with a desired set of data.
+type CARoot struct {
+	data                    map[string]string
+	configMapName           string
+	leaderElectionNamespace string
+
+	log    *logrus.Entry
+	client kubernetes.Interface
+
+	workqueue          workqueue.RateLimitingInterface
+	namespacesInformer cache.SharedInformer
+	configMapInformer  cache.SharedInformer
+
+	configMapLister corev1listers.ConfigMapLister
+	namespaceLister corev1listers.NamespaceLister
+}
+
+func NewCARootController(log *logrus.Entry, kubeOptions *options.KubeOptions,
+	leaderElectionNamepace string, configMapName string, data map[string]string) *CARoot {
+	return &CARoot{
+		data:                    data,
+		configMapName:           configMapName,
+		leaderElectionNamespace: leaderElectionNamepace,
+		log:                     log.WithField("module", "ca-root-controller"),
+		client:                  kubeOptions.KubeClient,
+		workqueue:               workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+}
+
+func (c *CARoot) Run(ctx context.Context, id string) {
+	rl := resourcelock.ConfigMapLock{
+		ConfigMapMeta: metav1.ObjectMeta{
+			Namespace: c.leaderElectionNamespace,
+			Name:      "cert-manager-istio-agent-election",
+		},
+		Client: c.client.CoreV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: id + "-cert-manager-istio-agent",
+		},
+	}
+
+	cancel := func() {}
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		ReleaseOnCancel: true,
+		Lock:            &rl,
+		LeaseDuration:   60 * time.Second,
+		RenewDeadline:   40 * time.Second,
+		RetryPeriod:     15 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				c.log.Info("acquired leader election")
+				ctx, cancel = context.WithCancel(ctx)
+				for {
+					if err := c.runController(ctx); err != nil {
+						c.log.Errorf("failed to run controller: %s", err)
+						time.Sleep(time.Second)
+						continue
+					}
+
+					break
+				}
+			},
+
+			OnStoppedLeading: func() {
+				c.log.Info("lost leader election")
+				cancel()
+			},
+		},
+	})
+}
+
+func (c *CARoot) runController(ctx context.Context) error {
```

Comment:
You didn't fancy using controller-runtime?

Reply:
This is mostly a copy of https://github.com/istio/istio/blob/master/security/pkg/k8s/controller/casecret.go

with a couple things changed